swarm.go

```diff
@@ -482,6 +491,14 @@ func (s *Swarm) Stop() error {
 	if s.pushSync != nil {
 		s.pushSync.Close()
 	}
+
+	if s.tags != nil {
+		err := s.stateStore.Put("tags", s.tags)
```

Comment:
Are we ok with ensuring tags persistence only if the swarm node shuts down cleanly.

For example, if the node is started with a loaded tags from disk, then it shuts down by panic or by operating system, the next swarm process will load the old tags which may already be processed and changed.

Reply:
that's a good question. i did not want to introduce a dependency for stateStore inside `chunk` package. I guess that another possibility would be to just have some pluggable function in `Tags` that would be set to the `stateStore.Put`, and we could just call this function periodically.

My concerns about this issue is that I don't want to incur a leveldb `Put` every time a counter changes on a `Tag`. If we choose to save on every counter change - then we can get rid of the `sync.Map` altogether since the data structure would be constantly persisted and maintain the in-memory copy becomes a burden.

The other thing is, that account for a panic is problematic - for example with pushsync you could have a data race where a receipt for a chunk would be received but the stateStore is not updated with the correct counters and a panic happens in between - even if you have persistence for the Tags you might end up in a corrupted state of the tag and it will never show up as `Done`(!).
thus removal of the chunk from the push index and the increment+persistence of the tag should be one atomic operation

Reply:
Yes, I agree about the problem with relation to push index. It would be good to have some integrity checks, maybe just to disregard tags persistence in case of the dirty flag, which could be set on swarm start and removed on clean start remove after successful tags store put. That way we would not slow down tags and protect in case of a hopefully rare crash (or os error, power loss...) situations.